iothub_client/src/iothubtransport_mqtt_common.c

```diff
@@ -1291,329 +1338,409 @@ static bool isSystemProperty(const char* tokenData)
-        else if (strcmp((const char*)&propName[nameLen - 2], CONTENT_ENCODING_PROPERTY) == 0)
+        break;
+
+        case MQTT_PROPERTY_TYPE_CONTENT_ENCODING:
         {
-            if (IoTHubMessage_SetContentEncodingSystemProperty(IoTHubMessage, propValue) != IOTHUB_MESSAGE_OK)
+            if (IoTHubMessage_SetContentEncodingSystemProperty(iotHubMessage, propValue) != IOTHUB_MESSAGE_OK)
             {
                 LogError("Failed to set IOTHUB_MESSAGE_HANDLE 'contentEncoding' property.");
                 result = MU_FAILURE;
             }
-            return result;
+            else
+            {
+                result = 0;
+            }
+        }
+        break;
+
+        default:
+        {
+            // This is an internal error in code as we should never have an unmapped property at this stage.
+            LogError("Unknown property type = %d to add to message", propertyType);
+            result = MU_FAILURE;
         }
     }
 
     return result;
 }
 
+
 //
-// extractMqttProperties parses the MQTT topic PUBLISH'd to this device/module, retrieves properties and fills out the
-// IOTHUB_MESSAGE_HANDLE which will ultimately be delivered to the application callback.
+// findMessagePropertyStart takes an MQTT topic PUBLISH'd as input and returns where message properties, if set, begin.
 //
-static int extractMqttProperties(IOTHUB_MESSAGE_HANDLE IoTHubMessage, const char* topic_name, bool urldecode)
+static const char* findMessagePropertyStart(PMQTTTRANSPORT_HANDLE_DATA transportData, const char* topic_name, IOTHUB_IDENTITY_TYPE type)
 {
-    int result;
-    STRING_TOKENIZER_HANDLE token = STRING_TOKENIZER_create_from_char(topic_name);
-    if (token != NULL)
+    const char *propertiesStart;
+
+    if (type == IOTHUB_TYPE_EVENT_QUEUE)
+    {
+        const char* input_queue_topic = STRING_c_str(transportData->topic_InputQueue);
+        // Substract one from queue length to reflect this having an extra # we subscribed to
+        propertiesStart = topic_name + (strlen(input_queue_topic) - 1);
+    }
+    else if (type == IOTHUB_TYPE_TELEMETRY)
+    {
+        const char* mqtt_message_queue_topic = STRING_c_str(transportData->topic_MqttMessage);
+        // Substract one from queue length to reflect this having an extra # we subscribed to
+        propertiesStart = topic_name + (strlen(mqtt_message_queue_topic) - 1);
+    }
+    else
+    {
+        LogError("ERROR: type %d is not expected", type);
+        propertiesStart = NULL;
+    }
+
+    return propertiesStart;
+}
+
+//
+// AddApplicationProperty adds the custom key/value property name from the incoming MQTT PUBLISH to the iotHubMessage
+// we will ultimately deliver to the application on its callback.
+//
+static int AddApplicationPropertyToMessage(MAP_HANDLE propertyMap, const char* propertyNameAndValue, size_t propertyNameLength, const char* propertyValue, bool auto_url_encode_decode)
```

Comment:
super-nit: You've used camelCase notation in all other static functions, but this one is starting with uppercase.